fabfile.py

```diff
@@ -46,258 +68,281 @@
-    channel.shutdown_write()
-    channel.shutdown_read()
-    remote_ret = channel.recv_exit_status()
-    if local_ret != 0 or remote_ret != 0:
-        raise Exception("remote_pipe failed. Local retcode: {0} Remote retcode: {1} output: {2}".format(local_ret, remote_ret, received))
-
-def relative(path):
-    """ Expresses a relative path as an absolute path using `env.real_fabfile`.
-        Paths are relative to the directory containing this file. """
-    directory = os.path.dirname(env.real_fabfile)
-    return os.path.join(directory, path)
-
-@contextmanager
-def esp_env():
-    with cd('%s/esp' % REMOTE_PROJECT_DIR):
-        with prefix('source %s/bin/activate' % REMOTE_VIRTUALENV_DIR):
-            yield
+@task
+def setup():
+    """
+    Perform initial configuration of a brand-new VM. May fail if run again.
+    """
+    env.setup_running = True
+
+    # Install Ubuntu packages, create a virtualenv and install Python packages.
+    # The script uses sudo to elevate as needed, so we can use run() here
+    # instead of sudo().
+    run(env.rbase + "esp/update_deps.sh --virtualenv=" + env.venv)
+
+    # Create and mount the encrypted partition (requires user input)
+    from fabtools import require
+    require.deb.package("cryptsetup")
+
+    print "***** "
+    print "***** Creating the encrypted partition for data storage."
+    print "***** Please choose a passphrase and type it at the three prompts."
+    print "***** "
+
+    sudo("cryptsetup luksFormat /dev/mapper/%s" % env.encvg)
+    sudo("cryptsetup luksOpen /dev/mapper/%s encrypted" % env.encvg)
+    sudo("mkfs.ext4 /dev/mapper/encrypted")
+    sudo("mkdir -p /mnt/encrypted")
+    sudo("mount /dev/mapper/encrypted /mnt/encrypted")
+
+    # Create the encrypted tablegroup in Postgres
+    sudo("chown -R postgres /mnt/encrypted")
+    psql("CREATE TABLESPACE encrypted LOCATION '/mnt/encrypted'")
+
+    # Automatically activate virtualenv. We rely on this so that we don't have
+    # to activate the virtualenv as part of every fab command.
+    files.append("~/.bash_login", "source %sbin/activate" % env.venv)
+
+    # Configure memcached item size
+    addendum = "\n# Item size limit\n-I 2M"
+    files.append("/etc/memcached.conf", addendum, use_sudo=True)
+    sudo("service memcached restart")
+
+    # Symlink media
+    if platform.system() == "Windows":
+        # Creating symlinks inside the guest will fail if the host is running
+        # Windows. We need to create the symlinks on the host, which requires
+        # administrator privileges. This script takes care of the whole process.
+        local(join(env.lbase, "deploy", "windows_symlink_media.bat"))
+    else:
+        with cd(env.rbase + "esp/public/media"):
+            with settings(warn_only=True):
+                run("ln -s -T default_images images")
+                run("ln -s -T default_styles styles")
+
+    sudo("touch /fab-setup-done")
+
+@runs_once
+def ensure_environment():
```

Comment:
As long as we have this anyway, it would be cool if it also checks that you've either run loaddb or emptydb. Right now, when you go straight from setup to runserver, you get a confusing error. (I checked because I was curious.)